### ISSUE TITLE
Referenced file is included even if its in the tsconfig exclude list

### ISSUE THREAD

**Issue by fis-cz** [Microsoft/TypeScript#15417]
**TypeScript Version:**  2.2.1 / Visual Studio Extension

Hello,

I would like to ask why referenced file is included to the compilation all the time even if it excluded in the tsconfig.

I am doing something like this:

I have 4 tsconfig files:

- tsconfig.json
   - extends tsconfig.Debug.json or tsconfig.Release.json based on VS solution configuration. Its auto-generated file written during the VS project pre-build task

```
rem *** Prepare tsconfig.json based on the solution configuration (Debug/Release)
echo // This file is generated automatically based on the Visual Studio Solution configuration. > "$(ProjectDir)tsconfig.json"
echo // To configure TypeScript transpiller for given solution configuration use the tsconfig.common.json, tsconfig.Debug.json and tsconfig.Release.json >> "$(ProjectDir)tsconfig.json"
echo { "compileOnSave": true, "extends" : "./tsconfig.$(ConfigurationName).json" } >> "$(ProjectDir)tsconfig.json"
```
and simply contains just:
```
// This file is generated automatically based on the Visual Studio Solution configuration. 
// To configure TypeScript transpiller for given solution configuration use the tsconfig.common.json, tsconfig.Debug.json and tsconfig.Release.json 
{ "compileOnSave": true, "extends" : "./tsconfig.Debug.json" } 
```

- tsconfig.common.json
   - common configuration independent on the VS solution configuration
- tsconfig.Debug.json
   - extends tsconfig.common.json and includes options related to debug version only (i.e. generating map files)
- tsconfig.Release.json
   - extends tsconfig.common.json and includes only options related to release version only (i.e. removeComments)

so I can do a conditional compilation of the project files, because I am also specifying in

tsconfig.Debug.json
```
   ...
   "exclude": [ ..., "/**/*.Release.ts" ]
```

tsconfig.Release.json
```
   ...
   "exclude": [..., "/**/*.Debug.ts"]
```
so basically all files with .Release.ts extension are excluded during the debug build and vice versa.

The problems comes when I have i.e. file with constants (this is just example) such as

constants.Debug.ts
```
   ...
   export const buildType: string = "Debug";
   ...
```
constants.Release.ts
```
   export const buildType: string = "Release";
   ...
```
which I am referencing in another file:

```
///<reference path="./constants.Debug.ts" />
///<reference path="./constants.Release.ts" />
...

```
**Expected behavior:**
I would expect that one of referenced files will not included to the project compilation (based on the build configuration) as it is excluded in the tsconfig

**Actual behavior:**
Both files are included so "Cannot redeclare block-scoped variable..." error is shown.

**Comment by mhegazy** [comment]
Please find more details in https://github.com/Microsoft/TypeScript/wiki/FAQ#why-is-a-file-in-the-exclude-list-still-picked-up-by-the-compiler

**Comment by fis-cz** [comment]
Thanks for response. Is there any way how to get around it? Because I need to reference the file in order it is included in the generated js before functions are used.

**Comment by mhegazy** [comment]
Would having different tsconfig.json files based on the setup be a possibility?

**Comment by fis-cz** [comment]
I am sorry, I don't see the idea. It probably would be possible but it also would be less comfortable for developers than the current solution.

**Comment by mhegazy** [comment]
[Configuration inheritance](http://www.typescriptlang.org/docs/handbook/tsconfig-json.html#configuration-inheritance-with-extends) can simplify the process.

**Comment by mhegazy** [comment]
otherwise, https://github.com/Microsoft/TypeScript/issues/3538 might be what you are looking for.

**Comment by fis-cz** [comment]
Thats exactly what I am doing. I inherit tsconfig.json from tsconfig.Debug.json which inherits tsconfig.Common.json. It works well until there is <reference path... inside of one of .ts files.

#3538 does not help at all.

**Comment by mhegazy** [comment]
do not mix references and tsconfig files. that would be my recommendation.

**Comment by fis-cz** [comment]
Anybody has an idea how to get around this problem?

**Comment by fis-cz** [comment]
Nobody was trying to solve this issue? Omg :)

**Comment by mhegazy** [comment]
Automatically closing this issue for housekeeping purposes. The issue labels indicate that it is unactionable at the moment or has already been addressed.
